server/sonar-server/src/main/java/org/sonar/server/permission/index/AuthorizationTypeSupport.java

```diff
@@ -0,0 +1,110 @@
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Optional;
+import java.util.Set;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.sonar.api.ce.ComputeEngineSide;
+import org.sonar.api.server.ServerSide;
+import org.sonar.server.es.NewIndex;
+import org.sonar.server.user.UserSession;
+
+import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+
+@ServerSide
+@ComputeEngineSide
+public class AuthorizationTypeSupport {
+
+  public static final String TYPE_AUTHORIZATION = "authorization";
+  public static final String FIELD_GROUP_IDS = "groupIds";
+  public static final String FIELD_GROUP_NAMES = "groupNames";
+  public static final String FIELD_USER_IDS = "userIds";
+  public static final String FIELD_UPDATED_AT = "updatedAt";
+
+  /**
+   * When true, then anybody can access to the project. In that case
+   * it's useless to store granted groups and users. The related
+   * fields are empty.
+   */
+  public static final String FIELD_ALLOW_ANYONE = "allowAnyone";
+
+  private final UserSession userSession;
+
+  public AuthorizationTypeSupport(UserSession userSession) {
+    this.userSession = userSession;
+  }
+
+  /**
+   * Creates a type that requires to verify that user has the read permission
+   * when searching for documents.
+   * It relies on a parent type named "authorization" that is automatically
+   * populated by {@link org.sonar.server.permission.index.PermissionIndexer}.
+   *
+   * Both types {@code typeName} and "authorization" are created. Documents
+   * must be created with _parent and _routing having the parent uuid as values.
+   *
+   * @see NewIndex#createTypeRequiringProjectAuthorization(String)
+   */
+  public static NewIndex.NewIndexType createTypeRequiringProjectAuthorization(NewIndex newIndex, String typeName) {
+    NewIndex.NewIndexType type = newIndex.createType(typeName);
+    type.setAttribute("_parent", ImmutableMap.of("type", TYPE_AUTHORIZATION));
+    type.setAttribute("_routing", ImmutableMap.of("required", "true"));
+
+    NewIndex.NewIndexType authType = newIndex.createType(TYPE_AUTHORIZATION);
+    authType.setAttribute("_routing", ImmutableMap.of("required", "true"));
+    authType.createDateTimeField(FIELD_UPDATED_AT);
+    authType.createLongField(FIELD_GROUP_IDS);
+    authType.stringFieldBuilder(FIELD_GROUP_NAMES).disableNorms().build();
+    authType.createLongField(FIELD_USER_IDS);
+    authType.createBooleanField(FIELD_ALLOW_ANYONE);
+    authType.setEnableSource(false);
+    return type;
+  }
+
+  /**
+   * Build a filter to restrict query to the documents on which
+   * user has read access.
+   */
+  public QueryBuilder createQueryFilter() {
+    Integer userId = userSession.getUserId();
+    Set<String> userGroupNames = userSession.getUserGroups();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this use of "getUserGroups"; it is deprecated. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)